### ISSUE TITLE
Stale activation collection

### ISSUE THREAD

**Issue by standsed** [dotnet/orleans#6921]
Orleans version 3.4.0

I'm having a lot of warnings
"ActivationCollector found an activation in a non Valid state. All activation inside the ActivationCollector should be in Valid state. Activation: {0}"

From within grain method I'm calling DeactivateOnIdle and judging from logs it's being deactivated successfully. I'm not sure that this guarantees that activation is removed from collection. Default collector config is used.

So far only strange thing which I have noticed is grainId format in logs
Usually:     cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785
In warning cf9576bb-0f56-4ca9-a6df-a100fb47cc9f-0x3163AF4B@1162e785

Any advice to track down the issue?

Debug level log for specific grain
```[
  {
    "@t": "2021-01-28T00:13:28.8995397Z",
    "@mt": "Silo {0} tries to lookup for {1}-->{2} ({3}-->{4})",
    "@l": "Debug",
    "0": "S10.233.108.240:10000:349462209",
    "1": "*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f",
    "2": "S10.233.108.240:10000:349462209",
    "3": 828616523,
    "4": -96743910,
    "SourceContext": "Orleans.Runtime.GrainDirectory.LocalGrainDirectory"
  },
  {
    "@t": "2021-01-28T00:13:28.8997583Z",
    "@mt": "Created {0} with GrainContext={1}",
    "@l": "Debug",
    "0": "ActivationTaskScheduler-127:Queued=0",
    "1": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785RandomPlacement State=Create]",
    "SourceContext": "Orleans.Runtime.Scheduler.ActivationTaskScheduler"
  },
  {
    "@t": "2021-01-28T00:13:28.8998457Z",
    "@mt": "CreateGrainInstance {0}{1}",
    "@l": "Debug",
    "0": "*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f",
    "1": "@1162e785",
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T00:13:28.8999250Z",
    "@mt": "About to call {1} grain's OnActivateAsync() method {0}",
    "@l": "Debug",
    "1": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785 #GrainType=Test.Service.Grains.TestMessageGrain Placement=RandomPlacement State=Activating]",
    "0": "Test.Service.Grains.TestMessageGrain",
    "EventId": {
      "Id": 100511
    },
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T00:13:28.9009535Z",
    "@mt": "Returned from calling {1} grain's OnActivateAsync() method {0}",
    "@l": "Debug",
    "1": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785 #GrainType=Test.Service.Grains.TestMessageGrain Placement=RandomPlacement State=Activating]",
    "0": "Test.Service.Grains.TestMessageGrain",
    "EventId": {
      "Id": 100512
    },
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T00:13:28.9011632Z",
    "@mt": "DeactivateActivationOnIdle: {0} {1}.",
    "0": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785 #GrainType=Test.Service.Grains.TestMessageGrain Placement=RandomPlacement State=Deactivating]",
    "1": "later when become idle",
    "EventId": {
      "Id": 100540
    },
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T00:13:28.9012566Z",
    "@mt": "About to call {1} grain's OnDeactivateAsync() method {0}",
    "@l": "Debug",
    "1": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785 #GrainType=Test.Service.Grains.TestMessageGrain Placement=RandomPlacement State=Deactivating]",
    "0": "Test.Service.Grains.TestMessageGrain",
    "EventId": {
      "Id": 100514
    },
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T00:13:28.9013066Z",
    "@mt": "Returned from calling {1} grain's OnDeactivateAsync() method {0}",
    "@l": "Debug",
    "1": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785 #GrainType=Test.Service.Grains.TestMessageGrain Placement=RandomPlacement State=Deactivating]",
    "0": "Test.Service.Grains.TestMessageGrain",
    "EventId": {
      "Id": 100515
    },
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T00:13:28.9017146Z",
    "@mt": "InitActivation is done: {0}",
    "@l": "Debug",
    "0": "S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f@1162e785",
    "SourceContext": "Orleans.Runtime.Catalog"
  },
  {
    "@t": "2021-01-28T02:14:10.4464231Z",
    "@mt": "ActivationCollector found an activation in a non Valid state. All activation inside the ActivationCollector should be in Valid state. Activation: {0}",
    "@l": "Warning",
    "0": "[Activation: S10.233.108.240:10000:349462209*grn/Test.Service.Grains.TestMessageGrain/cf9576bb-0f56-4ca9-a6df-a100fb47cc9f-0x3163AF4B@1162e785 #GrainType=Test.Service.Grains.TestMessageGrain Placement=RandomPlacement State=Invalid NonReentrancyQueueSize=0 EnqueuedOnDispatcher=0 InFlightCount=0 NumRunning=0 IdlenessTimeSpan=02:00:41.5451518 CollectionAgeLimit=02:00:00]",
    "EventId": {
      "Id": 100537
    },
    "SourceContext": "Orleans.Runtime.ActivationCollector"
  }
]```

**Comment by standsed** [comment]
Another thing which correlates with warnings is the sequence in which activation and deactivation actions complete

```

**Comment by ReubenBond** [comment]
Apologies for the slow response. I wouldn't rule out a race, but I wonder what's happening during deactivation which is causing this.
@benjaminpetit do you have any ideas?

**Comment by benjaminpetit** [comment]
I think the issue is that in [CallGrainActivate](https://github.com/dotnet/orleans/blob/3371650565343ea091976dacbfe16c76d225e7ba/src/Orleans.Runtime/Catalog/Catalog.cs#L1035), we start processing enqueued messages before we call [ActivationCollector.ScheduleCollection](https://github.com/dotnet/orleans/blob/3.4.2/src/Orleans.Runtime/Catalog/Catalog.cs#L633)

Which means that if a request, that calls `DeactivateOnIdle`, executes **before** the `ScheduleCollection` call, we will schedule the collection of an invalid Activation.

**Comment by benjaminpetit** [comment]
Fixed in #6974
